### PR TITLE
[ALST/Ulysses] Added ALST/Ulysses documentation

### PR DESCRIPTION
**Pull request by kashif**
# What does this PR do?

Added an ALST/Ulysses section to the distributed training doc with yaml config.

Needs: 
- [x] https://github.com/huggingface/accelerate/pull/3817 
- [x] https://github.com/huggingface/transformers/pull/41832
- [x] https://github.com/huggingface/transformers/pull/42444
- [ ] a new release with these PRs merged
 

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: docs/source/distributing_training.md [hunk 1]
```diff
@@ -185,6 +185,156 @@ These results show that **Context Parallelism (CP) scales effectively with more
 - [Hugging Face Blog: Enabling Long-Context Training with Sequence Parallelism in Axolotl](https://huggingface.co/blog/axolotl-ai-co/long-context-with-sequence-parallelism-in-axolotl)  
 - [Snowflake Engineering Blog: Arctic Long Sequence Training (ALST) — Scalable and Efficient Training for Multi-Million Token Sequences (Note that they use a different strategy)](https://www.snowflake.com/en/engineering-blog/arctic-long-sequence-training-multi-million-token-ai/)
 
+## ALST/Ulysses Sequence Parallelism (DeepSpeed)
```
Comment (sergiopaniego) [inline docs/source/distributing_training.md]: This section and the previous are related. We could make a section wrapping both of them and merging any repeated part. What do you think?

File: docs/source/distributing_training.md [hunk 2]
```diff
@@ -185,6 +185,156 @@ These results show that **Context Parallelism (CP) scales effectively with more
...
+
+parallelism_config = ParallelismConfig(
+    cp_backend="deepspeed",
+    cp_size=2,                # Number of GPUs for sequence parallelism
+    cp_handler=cp_handler,
+)
+
+# Training configuration
+training_args = SFTConfig(
+    max_length=4096,
```
Comment (sergiopaniego) [inline docs/source/distributing_training.md]: We need to define the max_lengt-seq_length twice?
Could we use the config in the yaml as in the previous example to remove the code above? For example:

```
deepspeed_config:
  ...
parallelism_config:
  parallelism_config_cp_size: 2
  ...
```

So a user can directly call the trainer and the structure is exactly the same as in the example above

File: docs/source/distributing_training.md [hunk 3]
```diff
@@ -185,6 +185,156 @@ These results show that **Context Parallelism (CP) scales effectively with more
...
+
+Then, launch your training script:
+
+```bash
+accelerate launch --config_file examples/accelerate_configs/alst_ulysses_4gpu.yaml train.py
+```
+
+### 2D Parallelism
+
+The 4 GPU configuration above automatically enables 2D parallelism by combining Data Parallelism (DP) with Context Parallelism (CP). The `dp_shard_size` is automatically calculated to distribute across available GPUs:
```
Comment (sergiopaniego) [inline docs/source/distributing_training.md]: We should add experiments here as the fsdp example

File: docs/source/distributing_training.md [hunk 4]
```diff
@@ -185,6 +185,156 @@ These results show that **Context Parallelism (CP) scales effectively with more
...
+
+```python
+import os
+
+num_gpus = int(os.environ.get("WORLD_SIZE", "1"))
+cp_size = 2
+dp_shard_size = num_gpus // cp_size  # Automatically calculated
+
+parallelism_config = ParallelismConfig(
```
Comment (sergiopaniego) [inline docs/source/distributing_training.md]: Do it via .yaml as in the other example to follow the same style?

File: docs/source/distributing_training.md [hunk 5]
```diff
@@ -185,6 +185,156 @@ These results show that **Context Parallelism (CP) scales effectively with more
...
+3. **Start with `cp_size=2`** before scaling to larger values
+4. **Use DeepSpeed ZeRO Stage 3** for large models
+5. **Combine with memory optimizations** like Liger kernels and gradient checkpointing
+
+### Complete Example
+
+Here's how to run ALST/Ulysses training using the built-in [`sft.py`](https://github.com/huggingface/trl/blob/main/trl/scripts/sft.py) script with 4 GPUs:
+
+```bash
+accelerate launch --config_file examples/accelerate_configs/alst_ulysses_4gpu.yaml \
```
Comment (sergiopaniego) [inline docs/source/distributing_training.md:338]: nice!

**Comment by SunMarc** [comment]
cc @stas00

File: docs/source/distributing_training.md [hunk 6]
```diff
@@ -185,6 +185,156 @@ These results show that **Context Parallelism (CP) scales effectively with more
 - [Hugging Face Blog: Enabling Long-Context Training with Sequence Parallelism in Axolotl](https://huggingface.co/blog/axolotl-ai-co/long-context-with-sequence-parallelism-in-axolotl)  
 - [Snowflake Engineering Blog: Arctic Long Sequence Training (ALST) — Scalable and Efficient Training for Multi-Million Token Sequences (Note that they use a different strategy)](https://www.snowflake.com/en/engineering-blog/arctic-long-sequence-training-multi-million-token-ai/)
 
+## ALST/Ulysses Sequence Parallelism (DeepSpeed)
+
+ALST (Arctic Long Sequence Training) / Ulysses is an alternative Context Parallelism implementation available through DeepSpeed. Unlike the FSDP2-based approach described above, ALST/Ulysses uses DeepSpeed's sequence parallelism infrastructure to split long sequences across GPUs.
+
+### Key Differences from FSDP2 Context Parallelism
```
Comment (stas00) [inline docs/source/distributing_training.md]: Kashif, do you want to mention somewhere that FSDP CP implements RSA/Ring Attention?

`FSDP2 Context Parallelism` doesn't tell the user what it is.

Probably in accelerate docs as well if it's not there?

Comment (stas00) [reply docs/source/distributing_training.md]: a good overview https://huggingface.co/blog/exploding-gradients/ulysses-ring-attention

Comment (stas00) [reply docs/source/distributing_training.md]: In particular I'd include this pros/cons of each method:

https://huggingface.co/blog/exploding-gradients/ulysses-ring-attention#ulysses-and-ring-attention

since it would help the user choose the most suitable method.

File: docs/source/distributing_training.md [hunk 7]
```diff
@@ -52,32 +52,70 @@ Example, these configurations are equivalent, and should yield the same results:
 > [!TIP]
 > Having one model per GPU can lead to high memory usage, which may not be feasible for large models or low-memory GPUs. In such cases, you can leverage [DeepSpeed](https://github.com/deepspeedai/DeepSpeed), which provides optimizations like model sharding, Zero Redundancy Optimizer, mixed precision training, and offloading to CPU or NVMe. Check out our [DeepSpeed Integration](deepspeed_integration) guide for more details.
 
-## Context Parallelism
+## Sequence Parallelism for Long Context Training
```
Comment (nmntz) [inline docs/source/distributing_training.md:55]: Maybe add note that we don't mean here the Sequence Parallelism which can be used in conjunction with Tensor Parallelism
This is strictly Context Parallelism
(for example TP=2 CP=2 will require you to have 4 gpus, whereas TP+SP=2 only needs 2)

File: docs/source/distributing_training.md [hunk 8]
```diff
@@ -52,32 +52,70 @@ Example, these configurations are equivalent, and should yield the same results:
...
-
-CP is particularly useful when:
+Sequence parallelism is particularly useful when:
 
 - You want to train with very long sequences (>32k tokens)
 - Single GPU memory is insufficient for your desired sequence length
 - You need to maintain sequence coherence across the full context
 
-### Requirements and Limitations
+### Available Implementations
```
Comment (nmntz) [inline docs/source/distributing_training.md:76]: since CP generalizes the concept of seqlen to microseqlen (within a single CP rank) / global seqlen, would be nice to add a note that seqlen previously denoted in TRL now refers to either microseqlen or globalseqlen (idk which one you're using)

File: examples/accelerate_configs/alst_ulysses_4gpu.yaml [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+main_training_function: main
+mixed_precision: bf16
+num_machines: 1
+num_processes: 4  # Total number of GPUs
+rdzv_backend: static
+same_network: true
+tpu_env: []
+tpu_use_cluster: false
+tpu_use_sudo: false
+use_cpu: false
```
Comment (sergiopaniego) [inline examples/accelerate_configs/alst_ulysses_4gpu.yaml:37]: Some params are missing?

```suggestion
use_cpu: false
parallelism_config:
  parallelism_config_dp_replicate_size: 1
  parallelism_config_dp_shard_size: 2  # Enables 2D parallelism with CP
  parallelism_config_tp_size: 1
  parallelism_config_cp_size: 2  # Context parallel size
  parallelism_config_cp_backend: deepspeed
  parallelism_config_cp_seq_length_is_variable: true
  parallelism_config_cp_attn_implementation: flash_attention_2
```

**Comment by SunMarc** [comment]
should we merge @kashif ?

**Comment by kashif** [comment]
sure let me do 1 final check @SunMarc
